File: guardian-alerts.ts
import fetch from "node-fetch";

export async function sendSlackAlert(message: string): Promise<boolean> {
  const url = process.env.GUARDIAN_SLACK_WEBHOOK_URL;
  if (!url) {
    console.log("[GuardianAlert] Slack webhook URL not configured.");
    return false;
  }
  try {
    const resp = await fetch(url, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ text: message }),
    });
    console.log("[GuardianAlert] Slack response:", resp.status);
    return resp.ok;
  } catch (e) {
    console.log("[GuardianAlert] Slack error:", e);
    return false;
  }
}

export async function sendMailgunEmail(
  subject: string,
  text: string,
  recipient: string
): Promise<boolean> {
  const domain = process.env.MAILGUN_DOMAIN;
  const key = process.env.MAILGUN_API_KEY;
  const sender =
    process.env.MAILGUN_FROM || (domain ? `guardian@${domain}` : "");
  if (!domain || !key || !sender) {
    console.log("[GuardianAlert] Mailgun config missing.");
    return false;
  }
  const url = `https://api.mailgun.net/v3/${domain}/messages`;
  const data = new URLSearchParams({
    from: sender,
    to: recipient,
    subject: subject,
    text: text,
  });

  try {
    const resp = await fetch(url, {
      method: "POST",
      headers: {
        Authorization: "Basic " + Buffer.from("api:" + key).toString("base64"),
      },
      body: data,
    });
    console.log("[GuardianAlert] Mailgun response:", resp.status);
    return resp.ok;
  } catch (e) {
    console.log("[GuardianAlert] Mailgun error:", e);
    return false;
  }
}

export async function sendSendgridEmail(
  subject: string,
  text: string,
  recipient: string
): Promise<boolean> {
  const apiKey = process.env.SENDGRID_API_KEY;
  const sender = process.env.SENDGRID_FROM;
  if (!apiKey || !sender) {
    console.log("[GuardianAlert] SendGrid config missing.");
    return false;
  }
  const url = "https://api.sendgrid.com/v3/mail/send";
  const data = {
    personalizations: [{ to: [{ email: recipient }] }],
    from: { email: sender },
    subject,
    content: [{ type: "text/plain", value: text }],
  };
  try {
    const resp = await fetch(url, {
      method: "POST",
      headers: {
        Authorization: `Bearer ${apiKey}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify(data),
    });
    console.log("[GuardianAlert] SendGrid response:", resp.status);
    return resp.ok;
  } catch (e) {
    console.log("[GuardianAlert] SendGrid error:", e);
    return false;
  }
}

export async function sendWebhookAlert(
  payload: object,
  url?: string
): Promise<boolean> {
  const targetUrl = url || process.env.GUARDIAN_ALERT_WEBHOOK_URL;
  if (!targetUrl) {
    console.log("[GuardianAlert] Webhook URL not set.");
    return false;
  }
  try {
    const resp = await fetch(targetUrl, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(payload),
    });
    console.log("[GuardianAlert] Webhook response:", resp.status);
    return resp.ok;
  } catch (e) {
    console.log("[GuardianAlert] Webhook error:", e);
    return false;
  }
}

type AlertKind = "slack" | "mailgun" | "sendgrid" | "webhook";

interface SendAlertOpts {
  kind: AlertKind;
  message?: string;
  subject?: string;
  recipient?: string;
  payload?: object;
  webhookUrl?: string;
}

export async function sendAlert(opts: SendAlertOpts): Promise<boolean> {
  const kind = opts.kind.toLowerCase();
  if (kind === "slack") {
    return sendSlackAlert(opts.message ?? "");
  } else if (kind === "mailgun") {
    if (opts.recipient && opts.subject && opts.message)
      return sendMailgunEmail(opts.subject, opts.message, opts.recipient);
  } else if (kind === "sendgrid") {
    if (opts.recipient && opts.subject && opts.message)
      return sendSendgridEmail(opts.subject, opts.message, opts.recipient);
  } else if (kind === "webhook") {
    return sendWebhookAlert(opts.payload || { text: opts.message }, opts.webhookUrl);
  }
  console.log("[GuardianAlert] Unhandled alert kind:", kind);
  return false;
}

File: guardian_alerts.py
import os
import requests
from typing import Optional, Dict

def send_slack_alert(message: str) -> bool:
    """Send a plain-text message to Slack via webhook."""
    url = os.getenv("GUARDIAN_SLACK_WEBHOOK_URL")
    if not url:
        print("[GuardianAlert] Slack webhook URL not configured.")
        return False
    payload = {"text": message}
    try:
        resp = requests.post(url, json=payload, timeout=10)
        print(f"[GuardianAlert] Slack response: {resp.status_code}")
        return resp.ok
    except Exception as e:
        print(f"[GuardianAlert] Slack error: {e}")
        return False

def send_mailgun_email(subject: str, text: str, recipient: str) -> bool:
    """Send alert email using Mailgun API."""
    domain = os.getenv("MAILGUN_DOMAIN")
    key = os.getenv("MAILGUN_API_KEY")
    sender = os.getenv("MAILGUN_FROM", f"guardian@{domain}" if domain else "")
    if not (domain and key and sender):
        print("[GuardianAlert] Mailgun config missing.")
        return False
    url = f"https://api.mailgun.net/v3/{domain}/messages"
    data = {
        "from": sender,
        "to": recipient,
        "subject": subject,
        "text": text
    }
    try:
        resp = requests.post(url, auth=("api", key), data=data, timeout=10)
        print(f"[GuardianAlert] Mailgun response: {resp.status_code}")
        return resp.ok
    except Exception as e:
        print(f"[GuardianAlert] Mailgun error: {e}")
        return False

def send_sendgrid_email(subject: str, text: str, recipient: str) -> bool:
    """Send alert email using SendGrid API."""
    api_key = os.getenv("SENDGRID_API_KEY")
    sender = os.getenv("SENDGRID_FROM")
    if not (api_key and sender):
        print("[GuardianAlert] SendGrid config missing.")
        return False
    url = "https://api.sendgrid.com/v3/mail/send"
    data = {
        "personalizations": [{"to": [{"email": recipient}]}],
        "from": {"email": sender},
        "subject": subject,
        "content": [{"type": "text/plain", "value": text}],
    }
    try:
        resp = requests.post(url, headers={
            "Authorization": f"Bearer {api_key}",
            "Content-Type": "application/json"
        }, json=data, timeout=10)
        print(f"[GuardianAlert] SendGrid response: {resp.status_code}")
        return resp.ok
    except Exception as e:
        print(f"[GuardianAlert] SendGrid error: {e}")
        return False

def send_webhook_alert(payload: Dict, url: Optional[str] = None) -> bool:
    """Send payload to a generic webhook (e.g., Discord, OpsGenie, custom)."""
    target_url = url or os.getenv("GUARDIAN_ALERT_WEBHOOK_URL")
    if not target_url:
        print("[GuardianAlert] Webhook URL not set.")
        return False
    try:
        resp = requests.post(target_url, json=payload, timeout=10)
        print(f"[GuardianAlert] Webhook response: {resp.status_code}")
        return resp.ok
    except Exception as e:
        print(f"[GuardianAlert] Webhook error: {e}")
        return False

def send_alert(
    kind: str,
    message: str = "",
    subject: str = "",
    recipient: str = "",
    payload: Optional[Dict] = None,
    webhook_url: str = "",
) -> bool:
    """Main dispatcher: kind = 'slack'|'mailgun'|'sendgrid'|'webhook'."""
    kind = kind.lower()
    if kind == "slack":
        return send_slack_alert(message)
    elif kind == "mailgun":
        if recipient and subject and message:
            return send_mailgun_email(subject, message, recipient)
    elif kind == "sendgrid":
        if recipient and subject and message:
            return send_sendgrid_email(subject, message, recipient)
    elif kind == "webhook":
        return send_webhook_alert(payload or {"text": message}, url=webhook_url)
    print(f"[GuardianAlert] Unhandled alert kind: {kind}")
    return False
